oak-run/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.jackrabbit.oak.indexversion;
+
+import org.apache.jackrabbit.oak.api.Type;
+import org.apache.jackrabbit.oak.plugins.index.search.IndexDefinition;
+import org.apache.jackrabbit.oak.plugins.index.search.spi.query.IndexName;
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.List;
+
+/*
+    Main operation of this class is to mark IndexName's with operations
+ */
+public class IndexVersionOperation {
+
+    enum Operation {
+        NOOP, DELETE_HIDDEN_AND_DISABLE, DELETE
+    }
+
+    private static final Logger log = LoggerFactory.getLogger(IndexVersionOperation.class);
+
+    private IndexName indexName;
+
+    public void setOperation(Operation operation) {
+        this.operation = operation;
+    }
+
+    public Operation getOperation() {
+        return this.operation;
+    }
+
+    public IndexName getIndexName() {
+        return this.indexName;
+    }
+
+    private Operation operation;
+
+    public IndexVersionOperation(IndexName indexName) {
+        this.indexName = indexName;
+        this.operation = Operation.NOOP;
+    }
+
+
+    public static List<IndexVersionOperation> applyOperation(NodeState indexDefParentNode,
+                                                             List<IndexName> indexNameObjectList, long threshold) {
+        int activeProductVersion = -1;
+        indexNameObjectList = getSortedVersionedIndexNameObjects(indexNameObjectList);
+        removeDisabledCustomIndexes(indexDefParentNode, indexNameObjectList);
+        List<IndexVersionOperation> indexVersionOperationList = new LinkedList<>();
+        for (int i = 0; i < indexNameObjectList.size(); i++) {
```

Comment:
I would use the enhanced for and avoid the number of calls to `indexNameObjectList.get(i)`